js/src/json/match.ts

```diff
@@ -0,0 +1,466 @@
+Make sure to evaluate each key the users ask you to evaluate separately. Assign the score
+for each key based on its own criteria - DO NOT convolute the scores of different keys.
+Also only evaluate the output vs. the reference output based on the criteria. DO NOT EVALUATE
+BASED ON ANYTHING ELSE. If the output does not match the reference output in some way that
+is not mentioned in the criteria that is not a problem and you should ignore those discrepancies.
+Only focus on finding discrepancies based on the criteria. If there is a None value being compared
+to a non-None value, you should assign a score of 0.`;
+
+const USER_PROMPT = `Please evaluate the accuracy of the following output keys according to these criteria:
+{rubric}
+<Outputs>
+{outputs}
+</Outputs>
+<Expected Outputs>
+{reference_outputs}
+</Expected Outputs>`;
+
+function _prepareParameters({
+    outputs,
+    referenceOutputs,
+    rubric,
+    excludeKeys,
+    useReasoning,
+    listMatchMode = 'same_elements'
+}: {
+    outputs: Record<string, any>[];
+    referenceOutputs: Record<string, any>[];
+    rubric: Record<string, string>;
+    excludeKeys: string[];
+    useReasoning: boolean;
+    listMatchMode: ListMatchMode;
+}): [any, any, any, Record<string, any>, string, boolean] {
+    const jsonSchema: Record<string, any> = {
+        type: 'object',
+        title: 'structured_match_score',
+        description: 'Scores measuring the accuracy of structured outputs',
+        properties: {},
+        required: [],
+        additionalProperties: false,
+    };
+
+    let scores: Record<string, any> = {};
+    let formattedRubric = '';
+    let useListReducer = false;
+    let processedOutputs: Record<string, any> = {};
+    let processedReferenceOutputs: Record<string, any> = {};
+
+    if (Array.isArray(outputs)) {
+        useListReducer = true;
+        if (!Array.isArray(referenceOutputs)) {
+            throw new Error('If outputs is a list, referenceOutputs must also be a list');
+        }
+
+        let outputsToUse: Record<string, any> = {};
+        let referenceOutputsToUse: Record<string, any> = {};
+
+        if (listMatchMode === 'ordered') {
+            outputs.forEach((output, i) => {
+                Object.entries(output).forEach(([key, value]) => {
+                    outputsToUse[`${key}_${i}`] = value;
+                });
+            });
+            referenceOutputs.forEach((refOutput, i) => {
+                Object.entries(refOutput).forEach(([key, value]) => {
+                    referenceOutputsToUse[`${key}_${i}`] = value;
+                });
+            });
+        } else if (listMatchMode === 'superset') {
+            const availableOutputs = Array.from(Array(outputs.length).keys());
+            const matchedReferences = new Set<number>();
+
+            referenceOutputs.forEach((refItem: any, i: number) => {
+                let bestMatchScore = -1;
+                let bestMatchIdx: number | null = null;
```

Comment:
nit, prefer `undefined` over `null` in js

```suggestion
                let bestMatchIdx: number | undefined;
```